crypto/modes/siv128.c

```diff
@@ -0,0 +1,361 @@
+/*
+ * Copyright 2017 The OpenSSL Project Authors. All Rights Reserved.
+ *
+ * Licensed under the OpenSSL license (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+#include <string.h>
+#include <stdlib.h>
+#include <openssl/crypto.h>
+#include <openssl/cmac.h>
+#include "modes_lcl.h"
+
+#ifndef OPENSSL_NO_SIV
+
+static const union {
+    long one;
+    char little;
+} is_endian = { 1 };
```

Comment:
is this initialization of a union guaranteed to work with all compilers?

Reply:
Yes. In C90, only one initializer can be used in a union initializer, and it will initialize the first member of the union. In C99, you can specify which field is initialized, but the default is the first field, and there can still only be one initializer.
This is also used in other places in OpenSSL (blake2, gcm128, xts128, chacha20, sha256). A good optimizing compiler will include only the correct endian-ness code.

Reply:
Speaking of which. You might have noticed that in other places this variable is never static. Reason for that is maximizing number of compilers that manage to eliminate dead code based on value. I mean you'll find compilers that will fail to eliminate dead code if it's static and succeed if it's not.

Reply:
In all (most?) other places, it's local to a function. I can put it locally into functions where needed, and make it non-static

Reply:
+1

Reply:
So, making it a static in function context would infer that storage is needed, where here, in file context, means that it's not used anywhere else, and can be optimized as needed for this compilation unit.

Reply:
@bernd-edlinger is that a request to put it inside the functions?

Reply:
yes, as automatic not static, makes sense for an optimizing compiler.

Reply:
Just in case. "Never static" in my previous remark does mean "always automatic". It's `const union {} is_endian` in each function, but not `static const`.

Reply:
ok; will do.